posix/dirent.c

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/dirent.h>
+#include <openenclave/corelibc/fcntl.h>
+#include <openenclave/corelibc/limits.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/trace.h>
+
+#define DIR_MAGIC 0x09180827
+
+struct _OE_DIR
+{
+    uint32_t magic;
+    int fd;
+    struct oe_dirent buf;
+};
+
+OE_DIR* oe_opendir_d(uint64_t devid, const char* pathname)
+{
+    OE_DIR* ret = NULL;
+    OE_DIR* dir = oe_calloc(1, sizeof(OE_DIR));
+    int fd = -1;
+    const int flags = OE_O_RDONLY | OE_O_DIRECTORY | OE_O_CLOEXEC;
+
+    if (!dir)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if ((fd = oe_open_d(devid, pathname, flags, 0)) < 0)
+        OE_RAISE_ERRNO_MSG(oe_errno, "pathname=%s", pathname);
+
+    dir->magic = DIR_MAGIC;
+    dir->fd = fd;
+
+    ret = dir;
+    dir = NULL;
+    fd = -1;
+
+done:
+
+    if (dir)
+        oe_free(dir);
+
+    if (fd >= 0)
+        oe_close(fd);
+
+    return ret;
+}
+
+OE_DIR* oe_opendir(const char* pathname)
+{
+    return oe_opendir_d(OE_DEVID_NONE, pathname);
+}
+
+struct oe_dirent* oe_readdir(OE_DIR* dir)
+{
+    struct oe_dirent* ret = NULL;
+    unsigned int count = (unsigned int)sizeof(struct oe_dirent);
+
+    if (!dir || dir->magic != DIR_MAGIC)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_getdents((unsigned int)dir->fd, &dir->buf, count) != (int)count)
+    {
+        if (oe_errno)
+            OE_RAISE_ERRNO(oe_errno);
+
+        goto done;
+    }
+
+    ret = &dir->buf;
+
+done:
+    return ret;
+}
+
+int oe_closedir(OE_DIR* dir)
+{
+    int ret = -1;
+
+    if (!dir || dir->magic != DIR_MAGIC)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    ret = oe_close(dir->fd);
+
+    oe_free(dir);
+
+done:
+
+    return ret;
+}
+
+void oe_rewinddir(OE_DIR* dir)
+{
+    if (dir && dir->magic == DIR_MAGIC)
+    {
+        oe_lseek(dir->fd, 0, OE_SEEK_SET);
+        dir->buf.d_off = 0;
+    }
+}
+
+int oe_getdents(unsigned int fd, struct oe_dirent* dirp, unsigned int count)
```

Comment:
>oe_getdents [](start = 4, length = 11)

For my understanding, do we really support `getdents` or are we mapping `readdir` semantics to `getdents`? The impl is not in this changeset, but `oe_dirent_t` only supports the standard fields, not the `linux_dirent` filetype.

If so, this is another non-standard behavior that needs to be explicitly documented.

Reply:
`readdir` isn't a system call in Linux, so you implement it using `getdents`. Here is an example in [musl.](https://github.com/ifduyue/musl/blob/79f653c6bc2881dd6855299c908a442f56cb7c2b/src/dirent/readdir.c) The main difference between the OE and `linux_dirent64` entry is that the Linux one doesn't have a max `d_name` size. However, since `NAME_MAX` in Linux is 255, the oe structure call always fit 1 entry.

Reply:
The implementation is consistent with the **getdents64** system call, not the **getdents** system call. The latter is obsolete and not used in modern Linux operating systems (not by MUSL), which explains why we haven't seen a problem with it yet. I removed the implementation of **getdents** and renamed **oe_getdents()** to **oe_getdents64()**. See the following reference for details about the **getdents64** system call.

http://man7.org/linux/man-pages/man2/getdents.2.html